crates/goose-mcp/src/google_drive/mod.rs

```diff
@@ -0,0 +1,583 @@
+                .build(
+                    hyper_rustls::HttpsConnectorBuilder::new()
+                        .with_native_roots()
+                        .unwrap()
+                        .https_or_http()
+                        .enable_http1()
+                        .build(),
+                );
+
+        DriveHub::new(client, auth)
+    }
+
+    pub async fn new() -> Self {
+        let drive = Self::google_auth().await;
+
+        // handle auth
+        let search_tool = Tool::new(
+            "search".to_string(),
+            indoc! {r#"
+                Search for files in google drive by name, given an input search query.
+            "#}
+            .to_string(),
+            json!({
+              "type": "object",
+              "properties": {
+              "query": {
+                  "type": "string",
+                  "description": "Search query",
+                  },
+              },
+              "required": ["query"],
+            }),
+        );
+
+        let read_tool = Tool::new(
+            "read".to_string(),
+            indoc! {r#"
+                Read a file from google drive using the file uri.
+                Optionally include base64 encoded images, false by default.
+            "#}
+            .to_string(),
+            json!({
+              "type": "object",
+              "properties": {
+                  "uri": {
+                      "type": "string",
+                      "description": "google drive uri of the file to read",
+                  },
+                  "includeImages": {
+                      "type": "boolean",
+                      "description": "Whether or not to include images as base64 encoded strings, defaults to false",
+                  }
+              },
+              "required": ["uri"],
+            }),
+        );
+
+        let instructions = indoc::formatdoc! {r#"
+            Google Drive MCP Server Instructions
+
+            ## Overview
+            The Google Drive MCP server provides two main tools for interacting with Google Drive files:
+            1. search - Find files in your Google Drive
+            2. read - Read file contents directly using a uri in the `gdrive:///uri` format
+
+            ## Available Tools
+
+            ### 1. Search Tool
+            Search for files in Google Drive, by name and ordered by most recently viewedByMeTime.
+            Returns: List of files with their names, MIME types, and IDs
+
+            ### 2. Read File Tool
+            Read a file's contents using its ID, and optionally include images as base64 encoded data.
+            The default is to exclude images, to include images set includeImages to true in the query.
+
+            Images take up a large amount of context, this should only be used if a
+            user explicity needs the image data.
+
+            ## File Format Handling
+            The server automatically handles different file types:
```

Comment:
Did you test if it can handle these different file types?

Reply:
i think the PNG one is missing which is okay. lets remove it from the instructions here

Reply:
ah yeah i'll need to add a note that for Google Sheets due to how the export works https://developers.google.com/drive/api/guides/ref-export-formats, it only exports the first sheet

Reply:
Added a note about the limitation and get this in response, when asking it to read a sheet.
```
0:"Note: This appears to be the first sheet of the spreadsheet, and due to technical limitations, I can only access the first sheet of Google Sheets documents.\n"
```